### PR TITLE
Add Tekton-owned Step, StepTemplate, and Sidecar

### PR DESCRIPTION
**Pull request by lbernick**
# Changes

This commit copies the Container struct from Kubernetes into Tekton into the
existing structs Task.Step.Container, Task.Sidecar.Container, and Task.StepTemplate.
This change breaks the Tekton Go libraries, but does not require any changes to the yaml
that users provide. This should not change anything for users who do not use the Go libraries.

The reason for this change is to have more control over our Task API. For example,
we may want to modify the Container fields we support (e.g. #4737), or change fields to string types
to allow them to be parameterized (e.g. #4080). This commit is the first step towards being able to make changes
to parts of the Task API embedded in Container.

Part 1 of #4737 

/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- n/a [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

# Release Notes

```release-note
[Change to Go libraries]: Task.Step, Task.StepTemplate, and Task.Sidecar use Container fields directly instead of embedding the Container struct
```

### CODE REVIEW

**Review by vdemeester — COMMENTED** [review]
Really quick review but few questions:
- Would it make sense to have two different struct for sidecars and steps ? (sidecars container would be step container + some things like readinessprobe, *probe …)
- I see `v1alpha1.Container`, any reason for changing code in `v1alpha1` ? (I guess it might be because we "share" the `Step` struct)

**Comment by lbernick** [comment]
> * Would it make sense to have two different struct for sidecars and steps ? (sidecars container would be step container + some things like readinessprobe, *probe …)

Yes that's definitely where I see this going but I didn't want to remove any fields of the Step container yet (might need a TEP). But I can create two identical structs for sidecars and steps that we can later change, what do you think?

> * I see `v1alpha1.Container`, any reason for changing code in `v1alpha1` ? (I guess it might be because we "share" the `Step` struct)

Yeah pretty much because we share the Step struct. I could update the v1alpha1 Step to just copy over all the fields of the v1beta1 step but then we have to keep them in sync. Open to doing that though.

**Comment by abayer** [comment]
fwiw, I'd lean towards copying over to v1alpha1 normally, but I'm not sure if it's worth it given that [TEP-0105](https://github.com/tektoncd/community/blob/main/teps/0105-remove-pipeline-v1alpha1-api.md) is starting to come together for actually removing v1alpha1. Given that, the additional work involved in copying over and keeping in sync for the time being may not be worth it.

**Comment by imjasonh** [comment]
> I didn't want to remove any fields of the Step container yet (might need a TEP)

Happy to discuss if folks disagree, but in general I don't think deprecating a "feature" like, e.g., `readinessProbe`s on a step container, should require a TEP. We should signal deprecations of these fields, wait 1+ release, and remove them. Especially so for things like `terminationMessage` which just silently don't do anything -- there's no way to use that correctly today.

If we think eventually we'd like to get to a spot where the `Step` and `Sidecar` types just have all the fields they reasonably could need and don't embed a `Container` type (even if it's owned by us), then I think we should just go straight to that and not incur the churn on ourselves and Go-API-consumers in having to account for two breaking type changes.

**Comment by lbernick** [comment]
> fwiw, I'd lean towards copying over to v1alpha1 normally, but I'm not sure if it's worth it given that [TEP-0105](https://github.com/tektoncd/community/blob/main/teps/0105-remove-pipeline-v1alpha1-api.md) is starting to come together for actually removing v1alpha1. Given that, the additional work involved in copying over and keeping in sync for the time being may not be worth it.

That's fair-- I wasn't sure if it would be more work to continue to alias v1alpha1 types to v1beta1 types, or to copy the old v1beta1 struct over to v1alpha1 so I wouldn't have to modify the v1alpha1 tests. I already did most of the grunt work though for pipelines -- when you're talking about the work to support this are you anticipating that this will be more difficult for the CLI team?

> > I didn't want to remove any fields of the Step container yet (might need a TEP)
> 
> Happy to discuss if folks disagree, but in general I don't think deprecating a "feature" like, e.g., `readinessProbe`s on a step container, should require a TEP. We should signal deprecations of these fields, wait 1+ release, and remove them. Especially so for things like `terminationMessage` which just silently don't do anything -- there's no way to use that correctly today.
> 
> If we think eventually we'd like to get to a spot where the `Step` and `Sidecar` types just have all the fields they reasonably could need and don't embed a `Container` type (even if it's owned by us), then I think we should just go straight to that and not incur the churn on ourselves and Go-API-consumers in having to account for two breaking type changes.

Point taken! I'd like to keep the scope of this PR to just swapping over the type definitions, and not deprecating anything. However, I created separate structs for Step and Sidecar containers to hopefully minimize churn. There's still a lot of copypasta so LMK if you'd prefer to see it done differently.

**Comment by lbernick** [comment]
tests pass! 💪

/assign @vdemeester 
/assign @imjasonh

**Comment by vdemeester** [comment]
> That's fair-- I wasn't sure if it would be more work to continue to alias v1alpha1 types to v1beta1 types, or to copy the old v1beta1 struct over to v1alpha1 so I wouldn't have to modify the v1alpha1 tests. I already did most of the grunt work though for pipelines -- when you're talking about the work to support this are you anticipating that this will be more difficult for the CLI team?
> 
> > >

We will remove support for `v1alpha1` in the cli very soon, if this push us to do it a bit early, I am more than fine with it 👼🏼.

> Especially so for things like `terminationMessage` which just silently don't do anything -- there's no way to use that correctly today.

Right, it's even more trickier than this. We use `terminationMessage` for Results, so as of today, this is very confusing to have / see in the spec.

> If we think eventually we'd like to get to a spot where the `Step` and `Sidecar` types just have all the fields they reasonably could need and don't embed a `Container` type (even if it's owned by us), then I think we should just go straight to that and not incur the churn on ourselves and Go-API-consumers in having to account for two breaking type changes.

That's a good point. I think we use the `Container` type today on both because one is not clearly wrapping the other one. With owning the struct, I think we may see that `Sidecar` has all the `Step` type and more. If that's the case, and we feel that's always gonna be the case, we could just merge `Step` and `Container` in the same struct as you suggest.

**Comment by imjasonh** [comment]
> Not sure which is worse, lots of copypasta or the potential for future additional breaking changes to the go libs (probably the latter?)

I'd personally go with copypasting in this case. I don't think there's much conceptual benefit in signalling "sidecars provide a superset of functionality of steps" and it sounds like we might not even believe that will continue to hold. It also means declaring the Go structs to be a little more awkward (`Steps: []Step{{Container: Container{Image: "..."}, Script: "..."}}` vs `Steps: []Step{{Image: "...", Script: "..."}}`).

I won't die on this hill though, if folks disagree.

Thanks for working on this, and for considering all these options as you do so 👍

**Comment by lbernick** [comment]
sounds good-- I think this PR should be good to go then!

**Comment by lbernick** [comment]
> > sounds good-- I think this PR should be good to go then!
> 
> I think with this latest change we should just drop the `StepContainer` and `SidecarContainer` structs, and move the fields directly into `Step` and `Sidecar`. WDYT?

Oh my bad, I understand now what you were suggesting. Will do, I'll comment here when I've made that change.

**Comment by lbernick** [comment]
I'm thinking of keeping StepContainer so it can be reused in StepTemplate, but copying the container fields into sidecar as well, thoughts?

File: pkg/apis/pipeline/v1beta1/task_types.go [hunk 1]
```diff
@@ -172,7 +172,7 @@ type Step struct {
 
 // Sidecar has nearly the same data structure as Step, consisting of a Container and an optional Script, but does not have the ability to timeout.
 type Sidecar struct {
-	corev1.Container `json:",inline"`
+	SidecarContainer `json:",inline"`
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1beta1/task_types.go:175]: The idea was to move all that is to not have this inlined/embedded struct but have the field directly in. So no need for `SidecarContainer`.

**Comment by lbernick** [comment]
> > I'm thinking of keeping StepContainer so it can be reused in StepTemplate, but copying the container fields into sidecar as well, thoughts?
> 
> Good question. Are there fields that make sense in a `Step` but not a `StepTemplate`, or vice versa? Can the type of `StepTemplate` just be `Step`?

I think "Name" wouldn't make sense in StepTemplate since [steps can't have the same name](https://github.com/tektoncd/pipeline/blob/2b8538214169ec1c3af14221f44ea59e6b1b81f7/pkg/apis/pipeline/v1beta1/task_validation.go#L176). In addition, "timeout" is not currently present in StepTemplate, but it could make sense. so maybe these fields should be directly copied into stepTemplate as well, thoughts?

**Comment by imjasonh** [comment]
> I think "Name" wouldn't make sense in StepTemplate since [steps can't have the same name](https://github.com/tektoncd/pipeline/blob/2b8538214169ec1c3af14221f44ea59e6b1b81f7/pkg/apis/pipeline/v1beta1/task_validation.go#L176). In addition, "timeout" is not currently present in StepTemplate, but it could make sense. so maybe these fields should be directly copied into stepTemplate as well, thoughts?

Thanks for looking into it. It sounds like we should just copy these between the types that need them. I'm not sure we should add `timeout` to StepTemplate in this PR, but it's something worth keeping in mind for a future change.

**Comment by lbernick** [comment]
> > I think "Name" wouldn't make sense in StepTemplate since [steps can't have the same name](https://github.com/tektoncd/pipeline/blob/2b8538214169ec1c3af14221f44ea59e6b1b81f7/pkg/apis/pipeline/v1beta1/task_validation.go#L176). In addition, "timeout" is not currently present in StepTemplate, but it could make sense. so maybe these fields should be directly copied into stepTemplate as well, thoughts?
> 
> Thanks for looking into it. It sounds like we should just copy these between the types that need them. I'm not sure we should add `timeout` to StepTemplate in this PR, but it's something worth keeping in mind for a future change.

Agreed! I don't want to change the user facing API at all with this PR-- just mentioning timeout as something that could be added in the future.

**Comment by lbernick** [comment]
@vdemeester @imjasonh I've addressed your feedback

**Review by imjasonh — COMMENTED** [review]
/lgtm

Thanks for this!

**Comment by lbernick** [comment]
> Thanks for this!

Happy to help!

/test pull-tekton-pipeline-alpha-integration-tests
/test pull-tekton-pipeline-integration-tests

**Review by vdemeester — APPROVED** [review]
/meow

**Pull request merged**